### PR TITLE
Change "early pywin32 import" warning back into an error

### PR DESCRIPTION
**Pull request by virtuald**
Reverts c901c4265c1bd41332396e6e7ec7f5290bd79bb4 and fixes the problem instead*

... I'm not currently able to run the tests reliably on Windows, so I'm doing this so I can watch them on appveyor. I'll comment here if I think it got fixed.

### CODE REVIEW

File: PyInstaller/compat.py [hunk 1]
```diff
@@ -870,10 +870,13 @@ def check_requirements():
         if 'win32api' in sys.modules or 'pywintypes' in sys.modules:
             # Users should never see this error; if it occurs, it means someone
```
Comment (xoviat) [inline PyInstaller/compat.py]: What about running check_requirements as soon as compat is imported? That would solve the problem and ensure that developers wouldn't need to worry about this in the future?

Comment (virtuald) [inline PyInstaller/compat.py]: I seem to remember trying this when I first made the changes, and it failed for some reason. I can't recall exactly why it failed, however.

Comment (xoviat) [inline PyInstaller/compat.py]: I cannot understand why this approach would succeed when the other approach would fail. Did you try simply putting `check_requirements()` at the end of the file?

**Comment by htgoebel** [comment]
What is the status of this pull-request? If nobody is working on it, please close it. and if this is na issue to be solved by another pull-request, please open an issue, thanks.

**Comment by virtuald** [comment]
It's still a problem that needs to be resolved, and I had forgotten about it. I'll try to look at it soon.

**Comment by htgoebel** [comment]
@virtuald May I kindly asking you to solve this issue. We already have a few tickets regarding this warning. So the warning creates irritation by the users and noise in the tickets. Thank you.

**Comment by virtuald** [comment]
I will try to resolve it this week, sorry for the delays!

**Comment by virtuald** [comment]
Working on this tonight, but tests on develop are failing for me on Windows, even on develop. Maybe I don't have something installed correctly.

File: PyInstaller/compat.py (hunk 1 above)
Comment (virtuald) [reply PyInstaller/compat.py]: The problem was circular imports of varying annoyance, so it seemed easier to not deal with them at the time. However, I've changed my mind and now I've changed it to force pywintypes et al to be retrieved from the compat module, which mostly solves the circular import issue.

**Comment by virtuald** [comment]
I still don't have my local environment working correctly, but I pushed something that probably works to AppVeyor -- we'll see if the tests pass. The problem is that it's building a stale commit, and I don't have any way of canceling the build.

If you enabled [rolling builds](https://github.com/appveyor/ci/issues/1529), then I think Appveyor would cancel builds when new commits are pushed, which could be a good thing.

File: PyInstaller/utils/win32/icon.py [hunk 2]
```diff
@@ -21,6 +21,8 @@
 except AttributeError:
     StringTypes = [ type("") ]
 
+from ...compat import win32api
```
Comment (virtuald) [inline PyInstaller/utils/win32/icon.py]: It seems that at one time pyinstaller supported not having win32api available. This is no longer the case, so I moved the import.

File: PyInstaller/utils/win32/winutils.py [hunk 3]
```diff
@@ -28,12 +26,8 @@ def get_windows_dir():
     """
     Return the Windows directory e.g. C:\\Windows.
     """
-    try:
-        import win32api
-    except ImportError:
```
Comment (virtuald) [inline PyInstaller/utils/win32/winutils.py]: Same thing here, if win32api isn't present then pyinstaller will exit, so no sense trying to detect this.

File: PyInstaller/utils/win32/winutils.py [hunk 4]
```diff
@@ -65,13 +56,14 @@ def extend_system_path(paths):
...
     """
+    from ... import compat
     old_PATH = compat.getenv('PATH', '')
     paths.append(old_PATH)
     new_PATH = os.pathsep.join(paths)
     compat.setenv('PATH', new_PATH)
 
 
-def import_pywin32_module(module_name):
+def import_pywin32_module(module_name, _is_venv=None):
```
Comment (virtuald) [inline PyInstaller/utils/win32/winutils.py]: Adding an internal-only parameter is necessary to prevent circular imports, since this function is called from `compat`

File: PyInstaller/utils/win32/winutils.py [hunk 5]
```diff
@@ -28,12 +26,8 @@ def get_windows_dir():
     """
     Return the Windows directory e.g. C:\\Windows.
     """
-    try:
-        import win32api
-    except ImportError:
-        windir = compat.getenv('SystemRoot', compat.getenv('WINDIR'))
-    else:
-        windir = win32api.GetWindowsDirectory()
+    from ... import compat
```
Comment (htgoebel) [inline PyInstaller/utils/win32/winutils.py]: Why importing this at the global level?

File: PyInstaller/utils/win32/winutils.py [hunk 6]
```diff
@@ -43,12 +37,9 @@ def get_system_path():
     """
     Return the path that Windows will search for dlls.
     """
+    from ... import compat
```
Comment (htgoebel) [inline PyInstaller/utils/win32/winutils.py]: Dito

File: PyInstaller/utils/win32/winutils.py [hunk 7]
```diff
@@ -65,13 +56,14 @@ def extend_system_path(paths):
 
     Some hooks might extend PATH where PyInstaller should look for dlls.
     """
+    from ... import compat
```
Comment (htgoebel) [inline PyInstaller/utils/win32/winutils.py]: dito.

File: PyInstaller/utils/win32/winutils.py (hunk 4 above)
Comment (htgoebel) [inline PyInstaller/utils/win32/winutils.py]: Please document this internal parameter in the doc-string.

File: PyInstaller/utils/win32/winutils.py [hunk 8]
```diff
@@ -114,7 +106,12 @@ def import_pywin32_module(module_name):
 
             # If isolated to a venv, the preferred site.getsitepackages()
             # function is unreliable. Fallback to searching "sys.path" instead.
-            if compat.is_venv:
+            if _is_venv is None:
```
Comment (htgoebel) [inline PyInstaller/utils/win32/winutils.py]: Why can't `compat.is_venv` be used?

File: PyInstaller/utils/win32/winutils.py [hunk 9]
```diff
@@ -114,7 +106,12 @@ def import_pywin32_module(module_name):
...
+                # prevents circular import
```
Comment (htgoebel) [inline PyInstaller/utils/win32/winutils.py]: Is this related to the import-statement? Then maybe it should read something like
"Don't import globally to prevents circular import"
and be put to all these nested import statements.

File: PyInstaller/compat.py [hunk 10]
```diff
@@ -198,6 +198,29 @@
...
+            pywintypes = winutils.import_pywin32_module('pywintypes', _is_venv=is_venv)
+            win32api = winutils.import_pywin32_module('win32api', _is_venv=is_venv)
+        except ImportError:
+            try:
+                from win32ctypes.pywin32 import pywintypes
+                from win32ctypes.pywin32 import win32api
+            except ImportError:
+                raise
+    except ImportError:
+        # Hack required to allow setup.py to work
```
Comment (htgoebel) [inline PyInstaller/compat.py]: Please add something like "This variable is set by setup.py" (this is not hard to guess, but not obvious ;-)

**Comment by htgoebel** [comment]
@virtuald Regarding the commit message: Maybe some of your inline remarks should go into there?

**Comment by virtuald** [comment]
Updated.

**Comment by htgoebel** [comment]
Thanks for the update :-)
